core/blockchain.go

```diff
@@ -427,20 +427,21 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 			bc.cacheConfig.TrieCleanRejournal = time.Minute
 		}
 		triedb := bc.stateCache.TrieDB()
-		bc.wg.Add(2)
-		go func() {
-			defer bc.wg.Done()
-			triedb.SaveCachePeriodically(bc.cacheConfig.TrieCleanJournal, bc.cacheConfig.TrieCleanRejournal, bc.quit)
-		}()
 		privatetriedb := bc.PrivateStateManager()
+		bc.wg.Add(1)
 		go func() {
 			defer bc.wg.Done()
-			privatetriedb.TrieDB().SaveCachePeriodically(bc.cacheConfig.PrivateTrieCleanJournal, bc.cacheConfig.TrieCleanRejournal, bc.quit)
+			bc.saveCachePeriodically(triedb, privatetriedb)
 		}()
 	}
 	return bc, nil
 }
 
+func (bc *BlockChain) saveCachePeriodically(triedb *trie.Database, privatetriedb mps.PrivateStateManager) {
+	triedb.SaveCachePeriodically(bc.cacheConfig.TrieCleanJournal, bc.cacheConfig.TrieCleanRejournal, bc.quit)
```

Comment:
Doesn't this mean that `triedb.SaveCachePeriodically` is executed and does not return until the `bc.quit` channel is closed and of course `privatetriedb.TrieDB().SaveCachePeriodically` is only invoked once the `bc.quit` channel has already been closed (in which case it is unlikely to do much)?

Reply:
It's a guess, and it's been a while I wanted to refactor this part.

Because in the issue, there is a complain that the data is not here for the private state.

Reply:
a part I made.
What version do you prefer?